core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
@junrao I agree with your points.  Thanks so much for the detailed explanation. I have updated the patch to include `LogDirFailureChannel` and an extra thread in `ReplicaManager`. The `LogManager.handleLogDirFailure()` will only be called by `ReplicaManager.handleLogDirFailure()`, which in turn will only be called by that thread. And the offline log dir name will be put into `LogDirFailureChannel` if there is new offline dir. The new thread will block waiting for new offline log dir.

I have also updated the patch to revert changes such as the ZkUtils in the LogManager constructor. I have reviewed the changes and the entire patches myself before uploading it. I think all comments have been addressed. Can you take another look at the patch? Thanks!